jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -2237,19 +2265,113 @@ private MetaData.Request extractMetaDataRequest(StreamFrame frame)
 
         private int reserveSlot(Slot slot, int streamId, Consumer<Throwable> fail)
         {
+            if (streamId < 0 || (streamId > 0 && !isLocalStream(streamId)))
+            {
+                fail.accept(new IllegalArgumentException("invalid stream id " + streamId));
+                return 0;
+            }
+
+            int maxTotal = getMaxTotalLocalStreams();
+
+            boolean created = false;
+            int reservedStreamId = 0;
             Throwable failure = null;
-            boolean reserved = false;
             try (AutoLock ignored = lock.lock())
             {
                 // SPEC: cannot create new streams after receiving a GOAWAY.
                 if (closed == CloseState.NOT_CLOSED)
                 {
-                    if (streamId <= 0)
+                    if (streamId == 0)
+                    {
+                        int total = totalLocalStreams.updateAndGet(v ->
+                        {
+                            if (v <= maxTotal)
+                                return v + 1;
+                            return v;
+                        });
+                        if (total <= maxTotal)
+                        {
+                            // Stream id generated internally.
+                            reservedStreamId = localStreamIds.getAndAdd(2);
+                            if (reservedStreamId > 0)
+                            {
+                                slots.offer(slot);
+                                created = true;
+                            }
+                            else
+                            {
+                                totalLocalStreams.decrementAndGet();
+                                failure = new IllegalStateException("max stream id exceeded");
+                            }
+                        }
+                        else
+                        {
+                            totalLocalStreams.decrementAndGet();
+                            failure = new IllegalStateException("max total streams exceeded");
+                        }
+                    }
+                    else
                     {
-                        streamId = localStreamIds.getAndAdd(2);
-                        reserved = true;
+                        // Stream id is given.
+                        while (true)
```

Comment:
This is a long maze of `if`s and `else`s in a `while` loop that partially duplicates the logic of the enclosing `if` branch, and is quite involved to follow.

Can't this be simplified and de-duplicated?